container-core/src/main/java/com/yahoo/container/core/config/ApplicationBundleLoader.java

```diff
@@ -48,54 +48,76 @@ public ApplicationBundleLoader(Osgi osgi, FileAcquirerBundleInstaller bundleInst
      */
     public synchronized Set<Bundle> useBundles(List<FileReference> newFileReferences) {
 
-        Set<FileReference> obsoleteReferences = getObsoleteFileReferences(newFileReferences);
-        Set<Bundle> bundlesToUninstall = getObsoleteBundles(obsoleteReferences);
+        obsoleteBundles = removeObsoleteBundles(newFileReferences);
+        Set<Bundle> bundlesToUninstall = new LinkedHashSet<>(obsoleteBundles.values());
         log.info("Bundles to schedule for uninstall: " + bundlesToUninstall);
 
         osgi.allowDuplicateBundles(bundlesToUninstall);
-        removeInactiveFileReferences(obsoleteReferences);
 
-        installBundles(newFileReferences);
+        bundlesFromNewGeneration = installBundles(newFileReferences);
         BundleStarter.startBundles(reference2Bundle.values());
         log.info(installedBundlesMessage());
 
         return bundlesToUninstall;
     }
 
-    private Set<FileReference> getObsoleteFileReferences(List<FileReference> newReferences) {
-        Set<FileReference> obsoleteReferences = new HashSet<>(reference2Bundle.keySet());
-        obsoleteReferences.removeAll(newReferences);
-        return obsoleteReferences;
+    /**
+     * Restores state from the previous application generation and returns the set of bundles that
+     * exclusively belongs to the latest (failed) application generation. Uninstalling must
+     * be done by the Deconstructor as they may still be needed by components from the failed gen.
+     */
+    public synchronized Collection<Bundle> revertToPreviousGeneration() {
+        reference2Bundle.putAll(obsoleteBundles);
+        bundlesFromNewGeneration.forEach(reference2Bundle::remove);
+        Collection<Bundle> ret = bundlesFromNewGeneration.values();
+
+        // No duplicate bundles should be allowed until the next call to useBundles.
+        osgi.allowDuplicateBundles(Set.of());
```

Comment:
This forcibly disallows duplicates, even though there could be running code from these bundles, as the new, failed graph isn't yet deconstructed. The normal flow is that a bundle is removed from the set of allowed duplicates (`allowedDuplicates` in `CollisionHook`) when that bundle is uninstalled. So can the `filter(...)` method of that class get called after we disallow duplicates _here_, but before the bundles are all cleaned out, and what happens in that case?

Reply:
If, instead, the argument was the set of new bundles, from the failed graph, which should be deconstructed, then these should be purged "in the normal fashion" as the revert was carried out, and code from both bundles would continue working normally until then ... no? (Assuming we keep the `CollisionHook` code which ensures the new allowed duplicates aren't duplicates of something that is _also_ an allowed duplicate, i.e., my temporary fix from earlier.)

Reply:
It's more complicated, I'll give you that. If what you propose here is safe, then it's the preferred choice :)

Reply:
Great catch. But now I don't see the problem of clearing the set of allowed duplicates, making it a 'set' operation instead of 'add'.  If you disagree, you are probably correct though.

Reply:
As long as we're guaranteed no new graph setup attempt is made before we complete deconstructing the failed graph, I think you're right; `SET` will be just fine, and it's cleaner.

Reply:
Ok, so if we have an ongoing revert, and a new deployment comes in, we have triplicated some of the bundles, and this isn't handled in any case, so as long as we don't also handle that, using a `SET` operation is fine. 
Deconstruction _is_ async, but having a new deployment concurrently with it seems unlikely.